src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -280,29 +295,173 @@ def process_input(conf):
             break
         upa, obj_info, genome_upa = task
 
-        # cfn points to the assembly file outside of the container
-        # get_assembly_as_fasta writes the file to /kb/module/workdir/tmp/<filename> inside the container.
-        # workdir is shared between the container and the external file system
-        # Any file path get_assembly_as_fasta returns will be relative to inside the container, and so is not useful for this script
+        upa_dir = os.path.join(conf.pth, upa)
+        metafile = os.path.join(upa_dir, f"{upa}.meta")
+        if not os.path.isdir(upa_dir) or not loader_helper.is_upa_info_complete(upa_dir):
+
+            # remove legacy upa_dir to avoid FileExistsError in hard link
+            if os.path.isdir(upa_dir):
+                shutil.rmtree(upa_dir)
+
+            # cfn points to the assembly file outside of the container
+            # get_assembly_as_fasta writes the file to /kb/module/workdir/tmp/<filename> inside the container.
+            # workdir is shared between the container and the external file system
+            # Any file path get_assembly_as_fasta returns will be relative to inside the container, and so is not useful for this script
+            cfn = os.path.join(conf.job_dir, "workdir/tmp", upa)
+            # upa file is downloaded to cfn
+            conf.asu.get_assembly_as_fasta({"ref": upa.replace("_", "/"), "filename": upa})
+
+            # each upa in output_dir as a separate directory
+            os.makedirs(upa_dir, exist_ok=True)
+
+            dst = os.path.join(upa_dir, f"{upa}.fa")
+            # Hard link .fa file from job_dir to output_dir in WS
+            os.link(cfn, dst)
+
+            # save meta file with relevant object_info
+            with open(metafile, "w", encoding="utf8") as json_file:
+                json.dump(_process_object_info(obj_info, genome_upa), json_file, indent=2)
+
+            print("Completed %s" % (upa))
+        else:
+            print(f"Skip downloading {upa} as it already exists")
+
+        if conf.retrieve_sample:
+            _download_sample_data(conf, genome_upa.replace("/", "_"), metafile)
```

Comment:
I would do the replace in the called method; make the arguments what people would naturally expect when they see `upa` in the args list

Reply:
I didn't get what you mean. `upa` is in format of `a_b_c` and `genome_up` is in format of `a/b/c`.

Reply:
I'm saying just pass `genome_upa` as is (e.g. `a/b/c`) and do the `replace` where needed in `_download_sample_data`. 

Also I see that in a sub method of `_download_sample_data` the opposite `replace` occurs, so those could be eliminated.

`a/b/c` is the standard form of an UPA and so we should pass that when it's reasonable to do so IMO

Reply:
right. I don't recall the history of `upa` and `genome_up`. But I think when we create file we want `a_b_c.xxx`. But when we write upa to a metadata file we want `a/b/c`. Thus there are a lot of `replace` here.

Reply:
probably need to clean up this. Because `upa` and `genome_up` in different format makes things unnecessary complicate.

Reply:
maybe
* pass everything as standard upa `a/b/c` and have a `to_filesafe_upa(upa)` method
* Make an UPA class where `str(new UPA(`a/b/c`)) == "a/b/c"`and `new UPA('a/b/c').filesafe() == `a_b_c`

Reply:
I just pass UPA in 'a/b/c' format for this PR. Unifying upa format across this class is way too beyond the scope.

Reply:
ok, would you mind making github issue to track it?

Reply:
👍 https://github.com/kbase/collections/issues/331

Reply:
:+1: